WPFSKillTree/SkillTreeFiles/SkillTree.cs

```diff
@@ -1171,62 +1158,60 @@ public void CrossAllHighlightedNodes()
-                return;
+        public static BuildUrlData DecodeUrl(string url, out HashSet<SkillNode> skillednodes)
+        {
+            int chartype;
+            int ascType;
+            return DecodeUrlProtected(url, out skillednodes, out chartype, out ascType);
+        }
 
-            url = Regex.Replace(url, @"\t| |\n|\r", "");
-            string s =
-                url.Substring(TreeAddress.Length + (url.StartsWith("https") ? 0 : -1))
-                    .Replace("-", "+")
-                    .Replace("_", "/");
-            byte[] decbuff = Convert.FromBase64String(s);
-            int i = BitConverter.ToInt32(new[] { decbuff[3], decbuff[2], decbuff[1], decbuff[0] }, 0);
-            byte b = decbuff[4];
-            long j = 0L;
-            byte asc = decbuff[5];
-            if (decbuff.Length >= 7)
-                j = decbuff[6];
-            var nodes = new List<ushort>();
-            for (int k = (i > 3 ? 7 : 6); k < decbuff.Length; k += 2)
-            {
-                byte[] dbff = { decbuff[k + 1], decbuff[k + 0] };
-                if (Skillnodes.Keys.Contains(BitConverter.ToUInt16(dbff, 0)))
-                    nodes.Add((BitConverter.ToUInt16(dbff, 0)));
-            }
-            chartype = b;
-            asctype = asc;
+        protected static BuildUrlData DecodeUrlProtected(string url, out HashSet<SkillNode> skillednodes, out int chartype, out int ascType)
+        {
+            BuildUrlData buildData = BuildConverter.GetUrlDeserializer(url).GetBuildData();
+
+            chartype = (byte)buildData.CharacterClassId;
+            ascType = (byte)buildData.AscendancyClassId;
+
+            string charName = CharName[chartype];
+            SkillNode startnode = Skillnodes.First(nd => nd.Value.Name.Equals(charName, StringComparison.InvariantCultureIgnoreCase)).Value;
+            skillednodes = new HashSet<SkillNode>{ startnode };
 
-            SkillNode startnode = Skillnodes.First(nd => nd.Value.Name.ToUpperInvariant() == CharName[b]).Value;
-            skillednodes.Add(startnode);
-            if (asc > 0)
+            if (ascType > 0)
             {
-                SkillNode ascNode = Skillnodes.First(nd => nd.Value.ascendancyName == _ascClasses.GetClassName(b, asc) && nd.Value.IsAscendancyStart).Value;
+                string ascendancyClass = AscendancyClasses.GetClassName(chartype, ascType);
+                SkillNode ascNode = AscRootNodeList.First(nd => nd.ascendancyName == ascendancyClass);
                 skillednodes.Add(ascNode);
             }
 
-            foreach (var node in nodes)
+            foreach (var nodeId in buildData.SkilledNodesIds)
             {
-                skillednodes.Add(Skillnodes[node]);
+                SkillNode node;
+                if (!Skillnodes.TryGetValue(nodeId, out node))
+                {
+                    throw new KeyNotFoundException(L10n.Message("The build you are trying to load contains unknown nodes."));
```

Comment:
Generally the node ids don't change much inbetween (unless there is a major rework). 
There isn't much we actually have to deal with since we use a hash it is O(1) to check if a node is there, and if it isn't move on.